### PR TITLE
PR: Add a finder widget to reuse a similar widget used in the Variable Explorer

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

File: spyder/widgets/helperwidgets.py [hunk 1]
```diff
@@ -308,48 +309,78 @@ def paintEvent(self, event):
...
             super(FinderLineEdit, self).keyPressEvent(event)
 
 
+class FinderWidget(QWidget):
+    sig_find_text = Signal(str)
+    sig_hide_finder_requested = Signal()
+
+    def __init__(self, parent, regex_base=None,
+                 key_filter_dict=None, find_on_change=False):
+        super(FinderWidget, self).__init__(parent)
```
Comment (ccordoba12) [inline spyder/widgets/helperwidgets.py:345]:
```suggestion
    def __init__(self, parent, regex_base=None, key_filter_dict=None,
                 find_on_change=False):
        super().__init__(parent)
```

File: spyder/widgets/helperwidgets.py [hunk 2]
```diff
@@ -308,48 +309,78 @@ def paintEvent(self, event):
...
+        # Parent is assumed to be a spyder widget
+        self.text_finder = FinderLineEdit(
+            self,
+            regex_base=regex_base,
+            key_filter_dict=key_filter_dict)
```
Comment (ccordoba12) [inline spyder/widgets/helperwidgets.py:350]:
```suggestion
            key_filter_dict=key_filter_dict
        )
```

**Review by ccordoba12 — COMMENTED** [review]
Thanks @impact27 for your work on this! I left a light review for you to start with.

**Comment by ccordoba12** [comment]
@impact27, I moved this to `master` because this refactoring doesn't fix bugs in Spyder 5, is significant and you mainly need it for PR #11186.

**Comment by impact27** [comment]
> @impact27, I moved this refactoring to `master` because it doesn't fix bugs in Spyder 5, it's significant and you mainly need it for PR #11186, which is also planned for Spyder 6.

@ccordoba12 I also need it for https://github.com/spyder-ide/spyder/pull/15372 . Should I move this one to `master` too?

**Comment by ccordoba12** [comment]
Yep, please do that as well.

File: spyder/plugins/variableexplorer/widgets/main_widget.py [hunk 3]
```diff
@@ -413,76 +405,44 @@ def on_section_conf_change(self, section):
...
-                callback=nsb.editor.set_regex,
-                main=nsb,
-            )
-
     def switch_widget(self, nsb, old_nsb):
         """
         Set the current NamespaceBrowser.
 
         This also setup the finder widget to work with the current
         NamespaceBrowser.
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/main_widget.py:413]:
```suggestion
        """Set the current NamespaceBrowser."""
```

The second part of this docstring is no longer correct.

File: spyder/plugins/variableexplorer/widgets/namespacebrowser.py [hunk 4]
```diff
@@ -118,21 +149,6 @@ def set_shellwidget(self, shellwidget):
...
-            self.text_finder.setText(self.last_find)
-        self.editor.finder = text_finder
-
-        return self.finder_is_visible
-
-    def save_finder_state(self, last_find, finder_visibility):
-        """Save last finder/search text input and finder visibility."""
-        if last_find and finder_visibility:
-            self.last_find = last_find
-        self.finder_is_visible = finder_visibility
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/namespacebrowser.py:134]: This is a loss of functionality, i.e. in the current implementation we're saving the last text that was introduced by the user in `text_finder` and we restore for the next search. That's very useful to switch back and forth between all variables and the ones you want to filter when running code several times.

So, please try to restore it in `FinderWidget`.

**Review by ccordoba12 — COMMENTED** [review]
@impact27, I think this is almost ready, except for the loss of functionality introduced by your refactoring (see below).

File: spyder/plugins/variableexplorer/widgets/namespacebrowser.py (hunk 4 above)
Comment (impact27) [reply spyder/plugins/variableexplorer/widgets/namespacebrowser.py:134]: Done

**Review by ccordoba12 — APPROVED** [review]
I tested your latest change manually and things are working as expected now.

Thanks @impact27!

**Pull request merged**